ui/apps/platform/src/reducers/publicConfig.ts

```diff
@@ -30,58 +35,76 @@ export const fetchPublicConfigThunk = () => {
 
 // Reducers
 
-const publicConfig = (state = {}, action) => {
-    if (action.type === types.FETCH_PUBLIC_CONFIG.SUCCESS) {
-        return isEqual(action.response, state) ? state : action.response;
-    }
-    return state;
-};
-
-const error = (state = null, action) => {
+const isLoadingPublicConfig: Reducer<boolean> = (state = true, action) => {
+    // Initialize true for edge case before authSagas call fetchUserRolePermissions action.
     switch (action.type) {
         case types.FETCH_PUBLIC_CONFIG.REQUEST:
-        case types.FETCH_PUBLIC_CONFIG.SUCCESS:
-            return null;
+            return true;
 
         case types.FETCH_PUBLIC_CONFIG.FAILURE:
-            return action.error;
+        case types.FETCH_PUBLIC_CONFIG.SUCCESS:
+            return false;
 
         default:
             return state;
     }
 };
 
-const isLoading = (state = true, action) => {
-    // Initialize true for edge case before authSagas call fetchUserRolePermissions action.
+const publicConfigInitialState: PublicConfig = {
+    footer: null,
+    header: null,
+    loginNotice: null,
+};
+
+const publicConfig: Reducer<PublicConfig, FetchingAction<{ response: PublicConfig }>> = (
+    state = publicConfigInitialState,
+    action
+) => {
+    if (action.type === types.FETCH_PUBLIC_CONFIG.SUCCESS) {
+        return isEqual(action.response, state) ? state : action.response;
+    }
+    return state;
+};
+
+const publicConfigError: Reducer<Error | null, ReturnType<FailureAction>> = (
+    state = null,
+    action
+) => {
     switch (action.type) {
         case types.FETCH_PUBLIC_CONFIG.REQUEST:
-            return true;
+        case types.FETCH_PUBLIC_CONFIG.SUCCESS:
+            return null;
 
         case types.FETCH_PUBLIC_CONFIG.FAILURE:
-        case types.FETCH_PUBLIC_CONFIG.SUCCESS:
-            return false;
+            return action.error;
 
         default:
             return state;
     }
 };
 
 const reducer = combineReducers({
+    isLoadingPublicConfig,
     publicConfig,
-    error,
-    isLoading,
+    publicConfigError,
 });
 
 // Selectors
 
-const getPublicConfig = (state) => state.publicConfig;
-const getPublicConfigError = (state) => state.error;
-const getIsLoadingPublicConfig = (state) => state.isLoading;
+type State = ReturnType<typeof reducer>;
```

Comment:
Even better than manually specifying the return type. 👍 

I tested this on my end for metadata I get a return type of `{ metadata: never }` due to the type I was using not actually being the exact type (an edge case I missed). Another win for defining types in terms of types. 😃

Reply:
Yes, thank you again for pointing me in a type-safer direction. There is room for improvement in follow ups.